python/pyspark/sql/types.py

```diff
@@ -368,8 +367,43 @@ def __init__(self, fields):
         >>> struct1 == struct2
         False
         """
-        assert all(isinstance(f, DataType) for f in fields), "fields should be a list of DataType"
-        self.fields = fields
+        if not fields:
+            self.fields = []
+        else:
+            self.fields = fields
+            assert all(isinstance(f, StructField) for f in fields),\
+                "fields should be a list of StructField"
+
+    def add(self, name_or_struct_field, data_type=None, nullable=True, metadata=None):
+        """
+        Construct a StructType by adding new elements to it to define the schema
+
+        >>> struct1 = StructType().add("f1", StringType(), True).add("f2", StringType(), True, None)
+        >>> struct2 = StructType([StructField("f1", StringType(), True),\
+         StructField("f2", StringType(), True, None)])
+        >>> struct1 == struct2
+        True
+        >>> struct1 = (StructType().add(StructField("f1", StringType(), True))
+        ...     .add(StructField("f2", StringType(), True, None)))
+        >>> struct2 = StructType([StructField("f1", StringType(), True),
+        ...     StructField("f2", StringType(), True, None)])
+        >>> struct1 == struct2
+        True
+
+        :param nameOrStructField: Either the name of the field or a StructField object
+        :param data_type: If present, the DataType of the StructField to create
+        :param nullable: Whether the field to add should be nullable (default True)
+        :param metadata: Any additional metadata (default None)
+        :return: a new updated StructType
+        """
+        if isinstance(name_or_struct_field, StructField):
+            self.fields.append(name_or_struct_field)
+            return self
+        else:
+            if isinstance(name_or_struct_field, str) and data_type is None:
+                raise ValueError("Must specify DataType if passing name of struct_field to create.")
+            self.fields.append(StructField(name_or_struct_field, data_type, nullable, metadata))
```

Comment:
Do we need to check the type of `data_type`? In scala, we support that `data_type` as a string, so it will be good that we also support it here. At least, we should check the type.